components/eventing-controller/controllers/subscription/nats/integration_test.go

```diff
@@ -0,0 +1,837 @@
+func TestUnavailableNATSServer(t *testing.T) {
+	natsPort := 4220
+	ctx := context.Background()
+	g := gomega.NewGomegaWithT(t)
+	ens := setupTestEnsemble(ctx, reconcilertesting.EventTypePrefix, g, natsPort, 7070)
+	defer ens.cancel()
+
+	subscription, subscriptionName := createSubscription(ctx, g, ens,
+		reconcilertesting.WithFilter("", uncleanEventType("")),
+		reconcilertesting.WithSinkURLFromSvc(ens.subscriberSvc),
+	)
+	testSubscriptionOnK8s(ctx, g, ens, subscription, subscriptionName,
+		reconcilertesting.HaveCondition(reconcilertesting.DefaultCondition("")),
+		reconcilertesting.HaveSubscriptionReady(),
+		reconcilertesting.HaveCleanEventTypes([]string{cleanEventType("")}),
+	)
+
+	ens.natsServer.Shutdown()
+	testSubscriptionOnK8s(ctx, g, ens, subscription, subscriptionName, reconcilertesting.HaveSubscriptionNotReady())
+
+	ens.natsServer = startNATS(natsPort)
+	testSubscriptionOnK8s(ctx, g, ens, subscription, subscriptionName, reconcilertesting.HaveSubscriptionReady())
+}
+
+//TestCreateSubscription tests if subscriptions get created properly.
+func TestCreateSubscription(t *testing.T) {
+	ctx := context.Background()
+	g := gomega.NewGomegaWithT(t)
+	ens := setupTestEnsemble(ctx, reconcilertesting.EventTypePrefix, g, 4225, 7075)
+	defer ens.cancel()
+
+	var testCases = []struct {
+		name               string
+		subscriptionOpts   []reconcilertesting.SubscriptionOpt
+		expect             expect
+		shouldTestDeletion bool
+	}{
+		{
+			name: "create and delete",
+			subscriptionOpts: []reconcilertesting.SubscriptionOpt{
+				reconcilertesting.WithFilter(reconcilertesting.EventSource, reconcilertesting.OrderCreatedEventTypeNotClean),
+				reconcilertesting.WithWebhookForNATS(),
+				reconcilertesting.WithSinkURLFromSvc(ens.subscriberSvc),
+			},
+			expect: expect{
+				k8sSubscription: []gomegatypes.GomegaMatcher{
+					reconcilertesting.HaveCondition(reconcilertesting.DefaultCondition("")),
+					reconcilertesting.HaveSubsConfiguration(configDefault(ens.defaultSubscriptionConfig.MaxInFlightMessages)),
+				},
+				natsSubscription: []gomegatypes.GomegaMatcher{
+					reconcilertesting.BeNotNil(),
+					reconcilertesting.BeValid(),
+					reconcilertesting.HaveSubject(reconcilertesting.OrderCreatedEventType),
+				},
+			},
+			shouldTestDeletion: true,
+		},
+
+		{
+			name: "filter with empty event type",
+			subscriptionOpts: []reconcilertesting.SubscriptionOpt{
+				reconcilertesting.WithFilter(reconcilertesting.EventSource, ""),
+				reconcilertesting.WithWebhookForNATS(),
+				reconcilertesting.WithSinkURLFromSvc(ens.subscriberSvc),
+			},
+			expect: expect{
+				k8sSubscription: []gomegatypes.GomegaMatcher{
+					reconcilertesting.HaveCondition(
+						eventingv1alpha1.MakeCondition(
+							eventingv1alpha1.ConditionSubscriptionActive,
+							eventingv1alpha1.ConditionReasonNATSSubscriptionNotActive,
+							v1.ConditionFalse, nats.ErrBadSubject.Error(),
+						),
```

Comment:
nit: There is a lot of nesting going on. We could refactor the condition to a function with name `NatsConditionWithErrorBadSubject` 🤔 

```suggestion
						reconcilertesting.NatsConditionWithErrorBadSubject
```

Which version do you prefer ?

Reply:
This is done for `conditionInvalidSink` as well in other tests in this file :)